backend/src/main/java/org/eclipse/tractusx/semantics/registry/security/OAuthSecurityConfig.java

```diff
@@ -31,53 +31,92 @@
+				.requestMatchers(HttpMethod.DELETE, "/**/shell-descriptors/**/submodel-descriptors/**").access("@authorizationEvaluator.hasRoleDeleteDigitalTwin()")
 
-                    // lookup
-                    // query endpoint is allowed for reader
-                    .requestMatchers( HttpMethod.POST, "/**/lookup/**/query/**" ).access( "@authorizationEvaluator.hasRoleViewDigitalTwin()" )
-                    // others are HTTP method based
-                    .requestMatchers( HttpMethod.GET, "/**/lookup/**" ).access( "@authorizationEvaluator.hasRoleViewDigitalTwin()" )
-                    .requestMatchers( HttpMethod.POST, "/**/lookup/**" ).access( "@authorizationEvaluator.hasRoleAddDigitalTwin()" )
-                    .requestMatchers( HttpMethod.PUT, "/**/lookup/**" ).access( "@authorizationEvaluator.hasRoleUpdateDigitalTwin()" )
-                    .requestMatchers( HttpMethod.DELETE, "/**/lookup/**" ).access( "@authorizationEvaluator.hasRoleDeleteDigitalTwin()" )
+				// lookup
+				// query endpoint is allowed for reader
+				.requestMatchers(HttpMethod.POST, "/**/lookup/**/query/**").access("@authorizationEvaluator.hasRoleViewDigitalTwin()")
+				// others are HTTP method based
+				.requestMatchers(HttpMethod.GET, "/**/lookup/**").access("@authorizationEvaluator.hasRoleViewDigitalTwin()").requestMatchers(HttpMethod.POST, "/**/lookup/**")
+				.access("@authorizationEvaluator.hasRoleAddDigitalTwin()").requestMatchers(HttpMethod.PUT, "/**/lookup/**")
+				.access("@authorizationEvaluator.hasRoleUpdateDigitalTwin()").requestMatchers(HttpMethod.DELETE, "/**/lookup/**")
+				.access("@authorizationEvaluator.hasRoleDeleteDigitalTwin()")
 
-                    //getDescription allowed for reader
-                    .requestMatchers( HttpMethod.GET, "/**/description" ).access( "@authorizationEvaluator.hasRoleViewDigitalTwin()" )
-              )
-              .csrf(CsrfConfigurer::disable)
-              .sessionManagement(sessionManagement -> sessionManagement.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
-              .oauth2ResourceServer(oauth2ResourceServerConfigurer -> oauth2ResourceServerConfigurer.jwt());
+				// getDescription allowed for reader
+				.requestMatchers(HttpMethod.GET, "/**/description").access("@authorizationEvaluator.hasRoleViewDigitalTwin()")).csrf(CsrfConfigurer::disable)
+				.sessionManagement(sessionManagement -> sessionManagement.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
+				.oauth2ResourceServer(oauth2ResourceServerConfigurer -> oauth2ResourceServerConfigurer.jwt());
 
-        return http.build();
-    }
+		return http.build();
+	}
 
-    @Bean
-    public AuthorizationEvaluator authorizationEvaluator(RegistryProperties registryProperties){
-        return new AuthorizationEvaluator(registryProperties.getIdm().getPublicClientId());
-    }
+	@Bean
+	public AuthorizationEvaluator authorizationEvaluator(final RegistryProperties registryProperties)
```

Comment:
The enum could have a simple method returning a new evaluator using the `registryProperties.getIdm()` as parameter. This way we could avoid the if statements. What do you think?

Reply:
I'm really not sure how this could work. Creating the evaluators via reflection is perhaps too dangerous in my opinion, because of different arguments. This would only work if both AuthorizationEvaluator and CognitoAuthorizationEvaluator have the same arguments in the constructor. Otherwise we always need the if...else.

Reply:
Sorry, I think we are not talking about the same thing. We don't need to use reflection, we can simply use inheritance and the fact that we can provide different implementations to the methods in the enum constant.
For example:

```java
public enum IdentityProvider {
   
   KEYCLOAK {
      @Override
      public AuthorizationEvaluator createAuthorizationEvaluator( RegistryProperties.Idm idmProperties ) {
         return new AuthorizationEvaluator( idmProperties.getPublicClientId() );
      }
   },
   
   COGNITO {
      @Override
      public AuthorizationEvaluator createAuthorizationEvaluator( RegistryProperties.Idm idmProperties ) {
         return new CognitoAuthorizationEvaluator( idmProperties );
      }
   };
   
   public abstract AuthorizationEvaluator createAuthorizationEvaluator( RegistryProperties.Idm idmProperties );
}
```